src/bananas.js

```diff
@@ -0,0 +1,185 @@
+        return transform(input, function(input) {
+            return xpathParser.parse(input).toHashtag();
+        });
+    }
+
+    /*
+     * hopefully robust parser that transforms any input (whether hashtag,
+     * xpath, or banana) into banana used by internal structures
+     */
+    function toBanana(input, xpathParser) {
+        if (!input) { return input; }
+        xpathParser = xpathParser || defaultParser;
+
+        var parsedBanana = transform(input);
+
+        try {
+            // we should really only accept bananas, but if they supply a nice
+            // looking xpath. let's turn them into pretty bananas
+            var parsed = xpathParser.parse(parsedBanana);
+            // success! now let's turn that input into bananas
+            parsed = xpathParser.parse(parsed.toHashtag());
+            return transformHashtags(parsed, xpathParser.models, function(input) {
+                return DELIMITER + input + DELIMITER;
+            });
+        } catch (err) {
+            // a bad xpath. let's just return the given input
+            return input;
+        }
+    }
+
+    /*
+     * This takes in a parsed hashtag (from xpathParser.parse) and transforms
+     * each hashtag into whatever is defined by transformFn
+     */
+    function transformHashtags(parsedHashtags, models, transformFn) {
+        var queue = [parsedHashtags],
+            EXPR = models.XPathInitialContextEnum.EXPR,
+            node, i, children, j, predicates;
+        while (queue.length > 0) {
+            node = queue.shift();
+            if (node instanceof models.XPathPathExpr) {
+                if (node.initial_context === EXPR) {
+                    queue.push(node.filter.expr);
+                    predicates = node.filter.predicates;
+                    for (i = 0; i < predicates.length; i++) {
+                        queue.push(predicates[i]);
+                    }
+                }
+            } else if (node instanceof models.HashtagExpr) {
+                (function() {
+                    var oldToHashtag = node.toHashtag;
+                    node.toHashtag = function() {
+                        return transformFn(oldToHashtag());
+                    };
+                })();
+            }
+            children = node.getChildren();
+            for (i = 0; i < children.length; i++) {
+                queue.push(children[i]);
+                if (children[i].predicates && children[i].predicates.length) {
+                    predicates = children[i].predicates;
+                    for (j = 0; j < predicates.length; j++) {
+                        queue.push(predicates[j]);
+                    }
+                }
+            }
+        }
+        return parsedHashtags.toHashtag();
+    }
+
+    /*
+     * transforms bananas based on transformFn
+     */
+    function transform(input, transformFn) {
+        if (!input) { return input; }
+        input = getSymbols(input);
```

Comment:
looks like input changes type from a string to an array. might want to use a new variable name to make that clear